### PR TITLE
Plugin to support libsvm format in datasets

### PR DESCRIPTION
**Pull request by ThmsLa**
LIBSVM format support, with two possible outputs:
- One feature per column
image
- All features in one JSON column
image

### CODE REVIEW

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                        limitedLogger.logV("Invalid token, skipping: %s %n", tokens[i]);
+                        continue;
+                    }
+
+                    int index = Integer.parseInt(pair[0]);
+                    double value = Double.parseDouble(pair[1]);
+                    
+                    // Adding a comma only if we're sure that another token has already been appended
+                    String comma = (features.length() > 3) ? "," : "";
+                    features.append(String.format(Locale.ROOT, "%s\"%d\":%f", comma, index, value));
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: No need for `format` with a `StringBuilder`

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        @Override
+        public void setOutputSchema(Schema schema) {
+        }
+
+        @Override
+        public void setWarningsContext(WarningsContext warningsContext) {
+        }
+    }
+
+    public static class LIBSVMFormatDetector implements CustomFormatSchemaDetector {
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: No implem?

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+import com.dataiku.dip.logging.LimitedLogFactory;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonElement;
+
+public class LIBSVMFormat implements CustomFormat {
+    public static final Pattern TOKEN_PATTERN = Pattern.compile("^(\\d+):([+-]?\\d+[.]?\\d*(?:[eE][+-]?\\d+)?)\\s+");
+    public static final Pattern LABEL_PATTERN = Pattern.compile("^([+-]?\\d+[.]?\\d*(?:[eE][+-]?\\d+)?)\\s+");
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: [citation needed] ;)

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    
+    /**
+     * Create a reader for a stream in the format
+     */
+    @Override
+    public CustomFormatInput getReader(JsonObject config, JsonObject pluginConfig) {
+        if (config != null) {
+            JsonElement tmpElt = config.get("max_features");
+            if (tmpElt != null && tmpElt.isJsonPrimitive())
+                maxFeatures = tmpElt.getAsInt();
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: Need brackets + could benefit from sanity check (>0 or at least >=0)

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+            
+            tmpElt = config.get("output_type");
+            if (tmpElt != null && tmpElt.isJsonPrimitive())
+                outputType = ExtractionMode.forName(tmpElt.getAsString());
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: Ditto (brackets)

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        /**
+         * Called if the schema is available (ie, dataset has been created)
+         */
+        @Override
+        public void setSchema(Schema schema, boolean allowExtraColumns) {
+        }
+
+        @Override
+        public void setWarningsContext(WarningsContext warnContext) {
+        }
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: Could use this to output warnings

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+                    }
+
+                    start = tokenMatcher.end();
+                    tokenMatcher.region(start, line.length());
+
+                    row.put(column, value);
+                }
+
+                if (!tokenMatcher.hitEnd()) {
+                    limitedLogger.logV("Invalid end-of-line at [%d,%d]: %s", lineNumber, start, line.substring(start));
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: The _token_ is invalid, not the end of the line (+ can be confusing if the error is around the middle, some users won't read the column position) → something like "Invalid token at line %d, column %d: '%s', ignoring the rest of that line".

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+                tokenMatcher.region(start, line.length());
+
+                while (tokenMatcher.find()) {
+                    String index = tokenMatcher.group(1);
+                    String value = tokenMatcher.group(2);
+
+                    Column column = columns.get(index);
+                    // Creating the newly found column if possible
+                    if (column == null) {
+                        if (columns.size() > maxFeatures) continue;
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: Infinite loop, no? The region remains the same. → Check in unit tests

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+                    limitedLogger.logV("Line %d is invalid (no label detected): %s", lineNumber, line);
+                    continue;
+                }
+                Row row = rf.row();
+
+                String label = labelMatcher.group(1);
+                row.put(labelColumn, label);
+
+                int start = labelMatcher.end();
+                tokenMatcher.region(start, line.length());
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: [Nitpick] Could factorize the code of the 2 methods up to this line

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+
+                // Creating the JSON using StringBuilder as it's a simple object
+                features.append("{");
+
+                while (tokenMatcher.find()) {
+                    String index = tokenMatcher.group(1);
+                    String value = tokenMatcher.group(2);
+
+                    if (features.length() > 3)
+                        features.append(',');
```
Comment (instanceofme) [inline libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: Brackets

File: libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java (hunk 9 above)
Comment (ThmsLa) [reply libsvm-format/java-formats/libsvm/src/com/dataiku/dss/formats/libsvm/LIBSVMFormat.java]: `start` is used in error messages

**Pull request merged**